### ISSUE TITLE
Add infrastructure for bidirectional streaming

### ISSUE THREAD

**Issue by Paebbels** [OSVVM/OSVVM-Common#4]
When implementing a verification component with bidirectional streaming like SPI. Some additions are needed in the `StreamTransactionPkg`.

1. Additional enumeration values in `StreamUnresolvedOperationType`:
   ```diff
     CHECK_BURST,
     TRY_CHECK_BURST,
   +  -- Bi-directional data transfer
   +  SEND_AND_GET,
   +  SEND_AND_GET_BURST,
   +
     MULTIPLE_DRIVER_DETECT  -- value used when multiple drivers are present
   ) ;
   ```
2. A new `Get` procedure with an input and output data parameter for bidirectional data exchange.  
   Because parameter modes do not distinguish subprogram overloads, a different name is needed.  
   In a proof-of-concept implementation, `Get2` is used.
   ```diff
     ------------------------------------------------------------
     procedure Get (
     ------------------------------------------------------------
       signal    TransactionRec  : inout StreamRecType ;
       variable  Data            : out   std_logic_vector ;
       variable  Param           : out   std_logic_vector ;
       constant  StatusMsgOn     : in    boolean := FALSE 
     ) ; 

   +  ------------------------------------------------------------
   +  procedure Get2 (
   +  ------------------------------------------------------------
   +    signal    TransactionRec  : inout StreamRecType ;
   +    variable  Data            : out   std_logic_vector ;
   +    constant  Param           : in    std_logic_vector ;
   +    constant  StatusMsgOn     : in    boolean := FALSE 
   +  ) ; 
   +
     ------------------------------------------------------------
     procedure Get (
     ------------------------------------------------------------
       signal    TransactionRec  : inout StreamRecType ;
       variable  Data            : out   std_logic_vector ;
       constant  StatusMsgOn     : in    boolean := FALSE 
     ) ; 
   ```
   Here the procedure body:
   ```diff
     ------------------------------------------------------------
     procedure Get (
     ------------------------------------------------------------
       signal    TransactionRec  : inout StreamRecType ;
       variable  Data            : out   std_logic_vector ;
       constant  StatusMsgOn     : in    boolean := FALSE 
     ) is 
     begin
       TransactionRec.Operation   <= GET ;
       TransactionRec.IntToModel  <= Data'length ;
       TransactionRec.BoolToModel <= StatusMsgOn ;     
       RequestTransaction(Rdy => TransactionRec.Rdy, Ack => TransactionRec.Ack) ; 
       Data  := SafeResize(TransactionRec.DataFromModel, Data'length) ; 
     end procedure Get ; 
  
   +  ------------------------------------------------------------
   +  procedure Get2 (
   +  ------------------------------------------------------------
   +    signal    TransactionRec  : inout StreamRecType ;
   +    variable  Data            : out   std_logic_vector ;
   +    constant  Param           : in    std_logic_vector ;
   +    constant  StatusMsgOn     : in    boolean := FALSE 
   +  ) is 
   +    variable LocalParam : std_logic_vector(TransactionRec.ParamToModel'length -1 downto 0) := (others => '-') ;
   +  begin
   +    LocalParam(Param'length-1 downto 0) := Param ; 
   +    LocalSend(TransactionRec, GET, (Data'range => 'U'), LocalParam, StatusMsgOn) ;
   +    Data  := SafeResize(TransactionRec.DataFromModel, Data'length) ; 
   +  end procedure Get2 ; 
   +	
     ------------------------------------------------------------
     procedure Get (
     ------------------------------------------------------------
       signal    TransactionRec  : inout StreamRecType ;
       variable  Data            : out   std_logic_vector ;
       variable  Param           : out   std_logic_vector ;
       constant  StatusMsgOn     : in    boolean := FALSE 
     ) is 
     begin
       Get(TransactionRec, Data, StatusMsgOn) ;
       Param := SafeResize(TransactionRec.ParamFromModel, Param'length) ; 
     end procedure Get ;  
   ```

**Comment by JimLewis** [comment]
Didn't you use the AddressBus package instead?

A SendGet would need to have iData and oData right?  As well as perhaps iParam and oParam?

**Comment by Paebbels** [comment]
The SPI has no addresses, it's a stream, hence the StreamTransactionPkg.

The `SendGet` subprograms are currently implemented directly in the SpiPkg. Maybe a generic variant can be found and put into Common.

**Comment by JimLewis** [comment]
Name Get2 is odd.

**Comment by Paebbels** [comment]
I'm don't like it either. I jutz used it to get the proof-of-concept to work. A final version - integrated into OSVVM - needs a better name.

As you can see in my mapping table, The SpiPkg use it's own names to distinguish the direction and what side is in control:
* control = transmit / receive
* submissive = provide / get

**Comment by JimLewis** [comment]
What I am seeing is that Get has different forms when there is a receiver vs a controller that is receiving the information.

Maybe that is what differentiates the two interfaces - an address bus has controller/manager and peripheral/subordinate/responder vs. the stream interface that has a TX and an RX.

**Comment by JimLewis** [comment]
Would it work to implement SPIReceive as a remapping to SendAndGet or is there something additional special about SPIReceive?

**Comment by JimLewis** [comment]
Updated with SendAndGet as proposed above.   It is on DEV branch.   Let me know what you think

**Comment by JimLewis** [comment]
Released in 2023.09
